reverse_engineering/helpers/connectionHelper.js

```diff
@@ -1,10 +1,95 @@
-function generateConnectionParams(connectionInfo) {}
+/**
+ * @typedef {import('../../shared/types').App} App
+ * @typedef {import('../../shared/types').Cluster} Cluster
+ * @typedef {import('../../shared/types').ConnectionInfo} ConnectionInfo
+ * @typedef {import('../../shared/types').ConnectionParams} ConnectionParams
+ */
 
-async function connect(connectionInfo) {}
+const clusterHelper = require('./clusterHelper');
+const { AUTH_TYPE, COUCHBASE_HOST_PREFIX, COUCHBASE_DEFAULT_KV_CONNECTION_PORT } = require('../../shared/constants');
 
-function close() {}
+let cluster = null;
+
+/**
+ * @param {{ connectionInfo: ConnectionInfo }} param0
+ * @returns {boolean}
+ */
+const isCloudStorage = ({ connectionInfo }) => {
+	return connectionInfo.host.startsWith(COUCHBASE_HOST_PREFIX.cloud);
+};
+
+/**
+ * @param {{ connectionInfo: ConnectionInfo }} param0
+ * @returns {string}
+ */
+const generateUrl = ({ connectionInfo }) => {
+	if (isCloudStorage({ connectionInfo })) {
+		return connectionInfo.host;
+	}
+
+	const keyValuePort = connectionInfo.kv_port || COUCHBASE_DEFAULT_KV_CONNECTION_PORT;
+
+	return `${COUCHBASE_HOST_PREFIX.selfHosted}${connectionInfo.host}:${keyValuePort}`;
+};
+
+/**
+ * @param {{ connectionInfo: ConnectionInfo }}
+ * @returns {ConnectionParams}
+ */
+const generateConnectionParams = ({ connectionInfo }) => {
+	if (connectionInfo.authType === AUTH_TYPE.securityCertificate) {
+		return {
+			url: generateUrl({ connectionInfo }),
+			options: {
+				authenticator: {
+					certificatePath: connectionInfo.security_certificate,
+					keyPath: connectionInfo.security_certificate_key,
+				},
+			},
+		};
+	}
+
+	return {
+		url: generateUrl({ connectionInfo }),
+		options: {
+			username: connectionInfo.couchbase_username,
+			password: connectionInfo.couchbase_password,
+		},
+	};
+};
+
+/**
+ * @param {{ connectionInfo: ConnectionInfo; app: App; }} param0
+ * @throws {Error}
+ * @returns {Promise<Cluster>}
+ */
+const connect = async ({ connectionInfo, app }) => {
+	if (cluster) {
+		return cluster;
+	}
+
+	const couchbase = await app.require('couchbase');
+	const { url, options } = generateConnectionParams({ connectionInfo });
+	cluster = await couchbase.connect(url, options);
```

Comment:
please double-check if this method may throw an Error, and if so - wrap that into the try/catch

Reply:
@serhii-filonenko  I have a question. Can this require('couchbase') cause an error? When Misha was working on couchbase upgrade he made a separate service for importing couchbase - services/nativeModules/loaders/importCouchbase.js
Here he used import. What do you think, may be it is worth using import?

Reply:
> please double-check if this method may throw an Error, and if so - wrap that into the try/catch

@chulanovskyi-bs There is no need to wrap it in a try-catch  since the function must return a cluster; if an error occurs, it will be caught at the RE API level

Reply:
@WilhelmWesser The solution was to install the patched package directly into the plugin, as we desided